### PR TITLE
New plugin spraymon 

### PR DESCRIPTION
**Pull request by BlindingRadiance**
Hi! I've made a new plugin `spraymon`. It's purpose is to detect possible spray attempts using win32k graphic objects by checking counters inside [W32PROCESS](https://www.geoffchappell.com/studies/windows/km/win32k/structs/processinfo/index.htm) structure. Since this structure is not really documented, symbol information is available only for Windows 7 and Windows 10 (>=1607). 
Checks are made when W32PROCESS information is being deleted from EPROCESS and at the end of the analysis.

### CODE REVIEW

File: src/plugins/spraymon/spraymon.cpp [hunk 1]
```diff
@@ -0,0 +1,328 @@
+/*********************IMPORTANT DRAKVUF LICENSE TERMS***********************
+*                                                                         *
+* DRAKVUF (C) 2014-2021 Tamas K Lengyel.                                  *
```
Comment (tklengyel) [inline src/plugins/spraymon/spraymon.cpp:3]: 2022

File: src/plugins/spraymon/spraymon.cpp [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+#include <json-c/json_object.h>
+
+#include "spraymon.h"
+#include "plugins/output_format.h"
+
+// #define PRINT_SPRAYMON(...) \
+//     do { \
+//         eprint_current_time(); \
+//         fprintf (stderr, __VA_ARGS__); \
+//     } while (0)
```
Comment (tklengyel) [inline src/plugins/spraymon/spraymon.cpp:116]: Remove this commented-out block if you don't use it anymore. You can also leave it in and just wrap it in an `ifdef DRAKVUF_DEBUG` block.

File: src/plugins/spraymon/spraymon.h [hunk 3]
```diff
@@ -0,0 +1,146 @@
+/*********************IMPORTANT DRAKVUF LICENSE TERMS***********************
+ *                                                                         *
+ * DRAKVUF (C) 2014-2021 Tamas K Lengyel.                                  *
```
Comment (tklengyel) [inline src/plugins/spraymon/spraymon.h:3]: 2022

**Comment by tklengyel** [comment]
Nice to see such quality submission on a first PR! :) LGTM, please fix the issues spotted by the CI and the couple comments I left and we can merge.

**Comment by BlindingRadiance** [comment]
Thanks!

**Comment by tklengyel** [comment]
@drakvuf-jenkins This is OK to test

File: src/plugins/spraymon/spraymon.cpp [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    {
+        PRINT_DEBUG("[SPRAYMON] Spraymon plugin supports only Windows 7 and Windows 10 (=>1607)\n");
+        do_final_analysis = false;
+        return;
+    }
+
+    if (!config->win32k_profile)
+    {
+        PRINT_DEBUG("[SPRAYMON] Win32k json profile required to run the plugin.\n");
+        throw -1;
```
Comment (tklengyel) [inline src/plugins/spraymon/spraymon.cpp:239]: Don't `throw` here if no win32k profile is set, just return gracefully.

File: src/plugins/spraymon/spraymon.cpp [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+    }
+    syscall = createSyscallHook("PsSetProcessWin32Process", &spraymon::hook_setwin32process_cb);
+    PRINT_DEBUG("[SPRAYMON]  PLUGIN STARTED\n");
+}
+
+spraymon::~spraymon()
+{
+
+}
```
Comment (BonusPlay) [inline src/plugins/spraymon/spraymon.cpp:272]: Do not declare empty dctor, it has certain implications.

File: src/plugins/spraymon/spraymon.h [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+struct spraymon_config
+{
+    const char* win32k_profile;
+    uint16_t gdi_threshold = 3000;
+    uint16_t usr_threshold = 3000;
+};
+
+class spraymon : public pluginex
+{
+public:
```
Comment (BonusPlay) [inline src/plugins/spraymon/spraymon.h:120]: no need to make everything public

File: src/plugins/spraymon/spraymon.cpp [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+
+    if (!read_counter(drakvuf, win32process + this->userhandlecountpeak, pid, usr_max_count))
+    {
+        PRINT_DEBUG("[SPRAYMON] Failed to read USER peak handle count\n");
+        return false;
+    }
+    return true;
+}
+
+void spraymon::compare(drakvuf_t drakvuf, uint16_t gdi_max_count, uint16_t usr_max_count, char* process_name, vmi_pid_t pid)
```
Comment (BonusPlay) [inline src/plugins/spraymon/spraymon.cpp:153]: misleading name as it doesn't really compare stuff, main reason is to output some log.

**Review by BonusPlay — CHANGES_REQUESTED** [review]
Nice code overall :)

**Comment by tklengyel** [comment]
@drakvuf-jenkins Retest this please

**Comment by BlindingRadiance** [comment]
Weird the error didn't show up on my local build.

**Pull request merged**